pkg/util/kubernetes/apiserver/util.go

```diff
@@ -61,3 +67,35 @@ func UnstructuredFromWPA(structIn *v1alpha1.WatermarkPodAutoscaler, unstructOut
 	unstructOut.SetUnstructuredContent(content)
 	return nil
 }
+
+func retryHandler(informers map[InformerName]cache.SharedInformer, name InformerName, config retryConfig) error {
+	nextTry := config.InitialRetryDelay
+	lastTry := false
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), nextTry)
+		start := time.Now()
+		if !cache.WaitForCacheSync(ctx.Done(), informers[name].HasSynced) {
+			log.Warnf("couldn't sync informer %s in %v", name, time.Now().Sub(start))
+			nextTry = config.method(nextTry, config)
+			config.RetryCount++
+			log.Warnf("Informers failed to sync, increasing timeout to %s", nextTry)
+			if lastTry {
+				cancel()
```

Comment:
Thanks for the suggestion, I think yours is right but I prefer to call the `cancel()` at the end of each iterations to avoid the `defer cancel `stacking up.  tbh, I don’t have a better solution

Reply:
> avoid the defer cancel stacking up

This is why this code is in a separate function: to create scope for defer. The loop calls this wrapper, and `cancel()` is called at the right time, once per iteration, without stacking up.

Another option is to assign result of `WaitForCacheSync` and cancel, then do the `if`:
```go
done := cache.WaitForCacheSync(ctx.Done(), hasSynched)
cancel()
if !done {
```

Reply:
Great, I just realise this is a very smart way to achieve the goal. Thanks 😊 . I will update this part of code